File: DailyTactics/Sources/Features/Tactics/TacticsViewModel.swift
import Foundation
import Observation

enum TacticsFeedbackState: Equatable {
    case instruction(message: String, systemImage: String)
    case error(message: String)
    case opponentMoving
    case opponentReply
    case incorrectMove
    case reviewing
    case puzzleComplete
    case trainingComplete
}

@MainActor
@Observable
final class TacticsViewModel {
    private var dataset: [Puzzle]
    private let dailyPuzzleCount: Int
    private(set) var puzzles: [Puzzle]
    private(set) var currentIndex: Int
    private(set) var session: PuzzleSession
    private(set) var selectedSquare: Square?
    private(set) var attemptedMove: ChessMove?
    private(set) var hintMove: ChessMove?
    private(set) var errorMessage: String?
    private var progress: PuzzleProgressStore?
    private let ratingStore: UserRatingStore
    private let ratingCalculator = PuzzleRatingCalculator()
    private var hadMistake = false
    private var ratingAppliedForPuzzle = false
    private var firstAttemptWasCorrect = false
    private var isAdvancing = false
    private(set) var userRating: Int
    private(set) var lastRatingDelta: Int?
    private(set) var isBoardFlipped: Bool = false
    private(set) var levelTransition: RatingLevel?

    /// Per-puzzle outcome for the current round, paralleling the rating
    /// assessment row. `nil` = not yet attempted.
    private(set) var results: [PuzzleOutcome?] = []

    init(dataset: [Puzzle] = Puzzle.loadBundled(), progress: PuzzleProgressStore? = nil, ratingStore: UserRatingStore = UserRatingStore(), dailyPuzzleCount: Int = 5, batchSize: Int? = nil) {
        let source = dataset.isEmpty ? Puzzle.samples : dataset
        let requestedCount = batchSize ?? dailyPuzzleCount
        let batch = Self.pickRandomBatch(from: source, count: requestedCount)
        self.dataset = source
        self.dailyPuzzleCount = requestedCount
        self.progress = progress
        self.ratingStore = ratingStore
        userRating = ratingStore.rating
        self.puzzles = batch
        results = Array(repeating: nil, count: batch.count)
        currentIndex = 0
        do {
            session = try PuzzleSession(puzzle: batch[0])
        } catch {
            preconditionFailure("Invalid bundled puzzle: \(error)")
        }
        orientBoardToPlayer()
    }

    func acknowledgeLevelTransition() {
        levelTransition = nil
    }

    func attachProgress(_ store: PuzzleProgressStore) {
        progress = store
    }

    /// Flip the board between the two playing perspectives.
    func toggleBoardFlip() {
        isBoardFlipped.toggle()
    }

    /// Orient the board so the player's own pieces are at the bottom. Called on
    /// every puzzle load so the user always starts from their perspective.
    private func orientBoardToPlayer() {
        isBoardFlipped = session.userColor == .black
    }

    // MARK: - Board state

    var position: [Square: Piece] { session.board.pieces }
    var lastMove: ChessMove? { session.lastMove }

    /// Position shown on the board. While a wrong move is being demonstrated, the
    /// moved piece is shown on its target square; clearing `attemptedMove` snaps it back.
    var displayedPosition: [Square: Piece] {
        guard let attempt = attemptedMove, let piece = position[attempt.from] else {
            return position
        }
        var preview = position
        preview.removeValue(forKey: attempt.from)
        preview[attempt.to] = piece
        return preview
    }

    var state: PuzzleSessionState { session.state }
    var playerColor: PieceColor { session.userColor }
    var headerTitle: String { state == .solved ? "Puzzle complete!" : "Your turn" }
    var headerSubtitle: String {
        "Find the best move for \(playerColor == .white ? "white" : "black")."
    }

    var feedbackState: TacticsFeedbackState {
        if errorMessage != nil {
            return .error(message: errorMessage ?? "")
        }
        if inReview && state != .solved {
            return .reviewing
        }

        switch state {
        case .waitingForMove:
            return .instruction(
                message: hintMove == nil ? "Find the winning move" : "Move the highlighted piece to the marked square",
                systemImage: hintMove == nil ? "scope" : "lightbulb.fill"
            )
        case .opponentMoving:
            return isReviewing ? .opponentReply : .opponentMoving
        case .incorrectMove:
            return .incorrectMove
        case .solved:
            return isBatchComplete ? .trainingComplete : .puzzleComplete
        }
    }

    // MARK: - Batch navigation

    var puzzleCount: Int { puzzles.count }
    var puzzleNumber: Int { currentIndex + 1 }
    var isLastPuzzle: Bool { currentIndex >= puzzles.count - 1 }
    var isBatchComplete: Bool { isLastPuzzle && session.state == .solved }

    func nextPuzzle() {
        guard !isAdvancing, currentIndex < puzzles.count - 1 else { return }
        isAdvancing = true
        let target = currentIndex + 1
        // A brief beat before the next puzzle appears so the transition reads
        // as deliberate rather than an instant snap. Re-entry is blocked until
        // the load completes so repeated taps can't skip puzzles.
        Task {
            try? await Task.sleep(for: .milliseconds(300))
            currentIndex = target
            loadPuzzle(at: target)
            isAdvancing = false
        }
    }

    /// Reshuffle a fresh random batch from the current dataset.
    func restartBatch() {
        reshuffleBatch()
    }

    /// Replace the dataset (e.g. after a rating-level transition imports a new
    /// tier) and start a fresh batch from it. A no-op when the new dataset is
    /// empty, so a failed import never leaves the user without puzzles.
    func reload(dataset: [Puzzle]) {
        guard !dataset.isEmpty else { return }
        self.dataset = dataset
        reshuffleBatch()
    }

    private func reshuffleBatch() {
        let available = dataset.filter { !(progress?.hasAttempted($0.id) ?? false) }
        let source = available.count >= dailyPuzzleCount ? available : dataset
        puzzles = Self.pickRandomBatch(from: source, count: dailyPuzzleCount)
        results = Array(repeating: nil, count: puzzles.count)
        currentIndex = 0
        loadPuzzle(at: 0)
    }

    // MARK: - Review stepping

    /// `<`/`>` are review-only: available once the puzzle is solved (or while
    /// scrubbing the line afterwards). Disabled during active play.
    var inReview: Bool { session.state == .solved || session.isReviewing }
    var canStepForward: Bool { inReview && session.canStepForward }
    var canStepBack: Bool { inReview && session.canStepBack }
    var hintEnabled: Bool {
        !inReview && (session.state == .waitingForMove || session.state == .incorrectMove)
    }
    var isReviewing: Bool { session.isReviewing }
    var currentMoveNumber: Int { session.currentMoveNumber }
    var totalUserMoves: Int { session.totalUserMoves }

    func stepForward() {
        guard canStepForward else { return }
        do {
            try session.stepForward()
            selectedSquare = nil
            attemptedMove = nil
            hintMove = nil
        } catch {
            errorMessage = "The next move could not be shown."
        }
    }

    /// Revealing a hint counts as giving up on this puzzle: it is scored
    /// immediately as a loss (rating decreases) and marked attempted, so a
    /// later clean solve can't recover the points.
    func requestHint() {
        guard hintEnabled, let expected = session.expectedMove else { return }
        hadMistake = true
        hintMove = expected
        applyHintPenalty()
    }

    /// Charge the rating for using a hint, once per puzzle. Idempotent so
    /// repeated taps don't stack penalties.
    private func applyHintPenalty() {
        guard !ratingAppliedForPuzzle else { return }
        ratingAppliedForPuzzle = true
        recordOutcome(.wrong, for: currentIndex)
        applySolveRating(solved: false)
        progress?.markAttempted(puzzles[currentIndex].id)
    }

    func stepBack() {
        guard canStepBack else { return }
        do {
            try session.stepBack()
            selectedSquare = nil
            attemptedMove = nil
            hintMove = nil
        } catch {
            errorMessage = "The previous move could not be shown."
        }
    }

    // MARK: - Session actions

    func start() {
        guard state == .opponentMoving, session.currentMoveIndex == 0 else { return }
        Task { await playOpponentMove() }
    }

    func select(_ square: Square) {
        guard !inReview,
              state == .waitingForMove || state == .incorrectMove else { return }
        attemptedMove = nil
        hintMove = nil

        if state == .incorrectMove {
            session.resumeAfterIncorrectMove()
        }

        if let selectedSquare {
            if selectedSquare == square {
                self.selectedSquare = nil
            } else if position[square]?.color == session.userColor {
                self.selectedSquare = square
            } else {
                attemptMove(from: selectedSquare, to: square)
            }
        } else if position[square]?.color == session.userColor {
            selectedSquare = square
        }
    }

    func restart() {
        loadPuzzle(at: currentIndex)
    }

    private static func pickRandomBatch(from dataset: [Puzzle], count: Int = 5) -> [Puzzle] {
        Array(dataset.shuffled().prefix(min(count, dataset.count)))
    }

    private func loadPuzzle(at index: Int) {
        do {
            session = try PuzzleSession(puzzle: puzzles[index])
            selectedSquare = nil
            hintMove = nil
            attemptedMove = nil
            errorMessage = nil
            hadMistake = false
            firstAttemptWasCorrect = false
            ratingAppliedForPuzzle = false
            lastRatingDelta = nil
            orientBoardToPlayer()
            Task { await playOpponentMove() }
        } catch {
            errorMessage = "The puzzle could not be loaded."
        }
    }

    private func attemptMove(from origin: Square, to target: Square) {
        hintMove = nil
        var move = ChessMove(from: origin, to: target)
        if session.moveNeedsPromotion(move) {
            move = ChessMove(from: origin, to: target, promotion: .queen)
        }
        // The expected move is trusted-legal (from the puzzle line), so accept
        // it even for special moves (castling, en passant) the basic legality
        // check does not model. Any other move must be basically legal.
        let isExpected = move == session.expectedMove
        guard isExpected || session.isLegalUserMove(move) else { return }

        let puzzleID = puzzles[currentIndex].id
        let isFirstAttempt = !(progress?.hasAttempted(puzzleID) ?? false)
        if isFirstAttempt {
            firstAttemptWasCorrect = isExpected
            progress?.markAttempted(puzzleID)
        }

        selectedSquare = nil
        do {
            try session.submitUserMove(move)
        } catch {
            errorMessage = "The move could not be applied."
            return
        }

        switch state {
        case .incorrectMove:
            // Wrong move: record it as a failure, then let the user retry.
            recordOutcome(.wrong, for: currentIndex)
            recordFailure()
            hadMistake = true
            attemptedMove = move
            let attempted = move
            Task {
                try? await Task.sleep(for: .milliseconds(550))
                guard !Task.isCancelled, attemptedMove == attempted else { return }
                attemptedMove = nil
            }
        case .solved:
            markCurrentSolved()
        case .opponentMoving:
            Task { await playOpponentMove() }
        default:
            break
        }
    }

    private func playOpponentMove() async {
        try? await Task.sleep(for: .milliseconds(450))
        guard !Task.isCancelled else { return }
        do {
            try session.applyOpponentMove()
        } catch {
            errorMessage = "The reply could not be applied."
            return
        }
        if session.state == .solved {
            markCurrentSolved()
        }
    }

    private func markCurrentSolved() {
        // Completion is recorded once regardless of the rating outcome: a hint
        // may have already adjusted the rating before the line is finished.
        progress?.markCompleted(puzzles[currentIndex].id)
        guard !ratingAppliedForPuzzle else { return }
        ratingAppliedForPuzzle = true
        recordOutcome(.correct, for: currentIndex)
        guard firstAttemptWasCorrect else { return }
        let cleanSolve = !hadMistake && hintMove == nil
        applySolveRating(solved: cleanSolve)
    }

    /// Applies an Elo-style rating change for the current puzzle, persists it,
    /// and surfaces a level-transition prompt when the user crosses a tier.
    private func applySolveRating(solved: Bool) {
        let puzzleRating = puzzles[currentIndex].rating ?? userRating
        let delta = ratingCalculator.change(
            userRating: userRating,
            puzzleRating: puzzleRating,
            solved: solved
        )
        userRating = ratingStore.apply(delta: delta)
        let newLevel = RatingLevel(rating: userRating)
        if newLevel != ratingStore.level {
            levelTransition = newLevel
            ratingStore.set(rating: userRating)
        }
        lastRatingDelta = delta
    }

    private func recordFailure() {
        guard let progress else { return }
        progress.markFailed(puzzles[currentIndex].id)
    }

    /// Mark a round outcome for the puzzle at `index`. Idempotent: the first
    /// recorded result (e.g. a wrong move) wins, so a later clean solve can't
    /// overwrite an earlier mistake.
    private func recordOutcome(_ outcome: PuzzleOutcome, for index: Int) {
        guard results.indices.contains(index), results[index] == nil else { return }
        results[index] = outcome
    }
}

File: android/app/src/main/java/com/dienbell/tactics/ui/tactics/TacticsViewModel.kt
package com.dienbell.tactics.ui.tactics

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dienbell.tactics.core.chess.ChessMove
import com.dienbell.tactics.core.chess.Piece
import com.dienbell.tactics.core.chess.PieceColor
import com.dienbell.tactics.core.chess.Square
import com.dienbell.tactics.core.puzzle.Puzzle
import com.dienbell.tactics.core.puzzle.PuzzleSession
import com.dienbell.tactics.core.puzzle.PuzzleSessionState
import com.dienbell.tactics.core.puzzle.PuzzleSessionState.INCORRECT_MOVE
import com.dienbell.tactics.core.puzzle.PuzzleSessionState.OPPONENT_MOVING
import com.dienbell.tactics.core.puzzle.PuzzleSessionState.SOLVED
import com.dienbell.tactics.core.puzzle.PuzzleSessionState.WAITING_FOR_MOVE
import com.dienbell.tactics.core.rating.PuzzleRatingCalculator
import com.dienbell.tactics.core.rating.RatingLevel
import com.dienbell.tactics.data.PuzzleProgressStore
import com.dienbell.tactics.data.UserRatingStore
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

sealed class TacticsFeedback {
    data class Instruction(val message: String) : TacticsFeedback()
    data class Error(val message: String) : TacticsFeedback()
    data object OpponentMoving : TacticsFeedback()
    data object OpponentReply : TacticsFeedback()
    data object IncorrectMove : TacticsFeedback()
    data object Reviewing : TacticsFeedback()
    data object PuzzleComplete : TacticsFeedback()
    data object TrainingComplete : TacticsFeedback()
}

data class TacticsUiState(
    val puzzleNumber: Int = 1,
    val puzzleCount: Int = 0,
    val displayedPosition: Map<Square, Piece> = emptyMap(),
    val lastMove: ChessMove? = null,
    val selectedSquare: Square? = null,
    val hintMove: ChessMove? = null,
    val playerColor: PieceColor = PieceColor.WHITE,
    val isBoardFlipped: Boolean = false,
    val sessionState: PuzzleSessionState = OPPONENT_MOVING,
    val inReview: Boolean = false,
    val isReviewing: Boolean = false,
    val canStepForward: Boolean = false,
    val canStepBack: Boolean = false,
    val hintEnabled: Boolean = false,
    val currentMoveNumber: Int = 0,
    val totalUserMoves: Int = 0,
    val userRating: Int = 1500,
    val lastRatingDelta: Int? = null,
    val results: List<PuzzleOutcome?> = emptyList(),
    val levelTransition: RatingLevel? = null,
    val feedback: TacticsFeedback = TacticsFeedback.Instruction("Find the winning move"),
) {
    val isBatchComplete: Boolean get() = puzzleNumber == puzzleCount && sessionState == SOLVED
}

/**
 * Coordinates a tactics round: session flow, hint (= immediate loss), rating,
 * review, level-transition reload, round results, and the next-puzzle delay.
 * Ported from the SwiftUI `TacticsViewModel`.
 */
class TacticsViewModel(
    private val dataset: List<Puzzle>,
    private val ratingStore: UserRatingStore,
    private val progressStore: PuzzleProgressStore? = null,
    private val calculator: PuzzleRatingCalculator = PuzzleRatingCalculator(),
    private val dailyPuzzleCount: Int = 5,
    private val importTier: (suspend (Int) -> Unit)? = null,
    private val scopedPuzzles: (suspend (Int) -> List<Puzzle>)? = null,
) : ViewModel() {

    private var source: List<Puzzle> = dataset.ifEmpty { Puzzle.SAMPLES }
    private var puzzles: List<Puzzle> = pickRandom(source, dailyPuzzleCount)
    private var currentIndex = 0
    private var session = PuzzleSession(puzzles.first())

    private var selectedSquare: Square? = null
    private var attemptedMove: ChessMove? = null
    private var hintMove: ChessMove? = null
    private var hadMistake = false
    private var ratingAppliedForPuzzle = false
    private var firstAttemptWasCorrect = false
    private var isAdvancing = false
    private var isBoardFlipped = false
    private var results: MutableList<PuzzleOutcome?> = MutableList(puzzles.size) { null }
    private val attemptedPuzzleIds = mutableSetOf<String>()

    private var userRating = ratingStore.rating()
    private var lastRatingDelta: Int? = null
    private var levelTransition: RatingLevel? = null

    private val _ui = MutableStateFlow(TacticsUiState())
    val ui: StateFlow<TacticsUiState> = _ui.asStateFlow()

    init {
        orientBoard()
        emit()
    }

    fun start() {
        if (session.state == OPPONENT_MOVING && session.currentMoveIndex == 0) {
            viewModelScope.launch { playOpponentMove() }
        }
    }

    fun toggleBoardFlip() { isBoardFlipped = !isBoardFlipped; emit() }

    fun acknowledgeLevelTransition() { levelTransition = null; emit() }

    /** Imports the new tier (if wired) and reloads the in-memory dataset from it. */
    fun confirmLevelTransition() {
        val rating = userRating
        viewModelScope.launch {
            importTier?.invoke(rating)
            scopedPuzzles?.invoke(rating)?.let { reload(it) }
            acknowledgeLevelTransition()
        }
    }

    fun select(square: Square) {
        if (inReview || (session.state != WAITING_FOR_MOVE && session.state != INCORRECT_MOVE)) return
        attemptedMove = null
        hintMove = null
        if (session.state == INCORRECT_MOVE) session.resumeAfterIncorrectMove()

        val selected = selectedSquare
        when {
            selected == null -> {
                if (session.board.pieces[square]?.color == session.userColor) selectedSquare = square
            }
            selected == square -> selectedSquare = null
            session.board.pieces[square]?.color == session.userColor -> selectedSquare = square
            else -> attemptMove(from = selected, to = square)
        }
        emit()
    }

    fun requestHint() {
        if (!hintEnabled) return
        val expected = session.expectedMove ?: return
        hadMistake = true
        hintMove = expected
        applyHintPenalty()
        emit()
    }

    fun stepForward() {
        if (!canStepForward) return
        runCatching { session.stepForward() }
            .onFailure { /* keep board; error is non-fatal for review */ }
        clearSelection()
        emit()
    }

    fun stepBack() {
        if (!canStepBack) return
        runCatching { session.stepBack() }.onFailure {}
        clearSelection()
        emit()
    }

    fun nextPuzzle() {
        if (isAdvancing || currentIndex >= puzzles.size - 1) return
        isAdvancing = true
        val target = currentIndex + 1
        // A brief beat so the transition reads as deliberate, not an instant snap.
        viewModelScope.launch {
            delay(NEXT_PUZZLE_DELAY_MS)
            currentIndex = target
            loadPuzzle(target)
            isAdvancing = false
            emit()
        }
    }

    fun restartBatch() { reshuffleBatch() }

    fun restart() { loadPuzzle(currentIndex); emit() }

    fun reload(dataset: List<Puzzle>) {
        if (dataset.isEmpty()) return
        source = dataset
        reshuffleBatch()
    }

    // region Internals

    private fun reshuffleBatch() {
        val available = source.filter { it.id !in attemptedPuzzleIds }
        val pool = if (available.size >= dailyPuzzleCount) available else source
        puzzles = pickRandom(pool, dailyPuzzleCount)
        results = MutableList(puzzles.size) { null }
        currentIndex = 0
        loadPuzzle(0)
        emit()
    }

    private fun loadPuzzle(index: Int) {
        session = PuzzleSession(puzzles[index])
        selectedSquare = null
        hintMove = null
        attemptedMove = null
        hadMistake = false
        firstAttemptWasCorrect = false
        ratingAppliedForPuzzle = false
        lastRatingDelta = null
        orientBoard()
        viewModelScope.launch { playOpponentMove() }
    }

    private fun attemptMove(from: Square, to: Square) {
        hintMove = null
        var move = ChessMove(from, to)
        if (session.moveNeedsPromotion(move)) move = ChessMove(from, to, com.dienbell.tactics.core.chess.PieceKind.QUEEN)

        val isExpected = move == session.expectedMove
        if (!isExpected && !session.isLegalUserMove(move)) return

        val puzzleId = puzzles[currentIndex].id
        val isFirstAttempt = puzzleId !in attemptedPuzzleIds
        if (isFirstAttempt) {
            firstAttemptWasCorrect = isExpected
            attemptedPuzzleIds.add(puzzleId)
            viewModelScope.launch { progressStore?.markAttempted(puzzleId) }
        }
        selectedSquare = null

        session.submitUserMove(move)

        when (session.state) {
            INCORRECT_MOVE -> {
                recordOutcome(PuzzleOutcome.WRONG, currentIndex)
                hadMistake = true
                attemptedMove = move
                viewModelScope.launch { progressStore?.markFailed(puzzleId) }
                viewModelScope.launch {
                    delay(WRONG_MOVE_PREVIEW_MS)
                    if (attemptedMove == move) attemptedMove = null
                    emit()
                }
            }
            SOLVED -> markCurrentSolved()
            OPPONENT_MOVING -> viewModelScope.launch { playOpponentMove() }
            else -> Unit
        }
    }

    private suspend fun playOpponentMove() {
        delay(OPPONENT_REPLY_MS)
        session.applyOpponentMove()
        if (session.state == SOLVED) markCurrentSolved()
        emit()
    }

    private fun markCurrentSolved() {
        if (ratingAppliedForPuzzle) return
        ratingAppliedForPuzzle = true
        viewModelScope.launch { progressStore?.markCompleted(puzzles[currentIndex].id) }
        recordOutcome(PuzzleOutcome.CORRECT, currentIndex)
        if (!firstAttemptWasCorrect) return
        val puzzleRating = puzzles[currentIndex].rating ?: userRating
        val cleanSolve = !hadMistake && hintMove == null
        applySolveRating(solved = cleanSolve, puzzleRating = puzzleRating)
    }

    /** A hint is a give-up: score an immediate loss, mark attempted. Idempotent. */
    private fun applyHintPenalty() {
        if (ratingAppliedForPuzzle) return
        ratingAppliedForPuzzle = true
        recordOutcome(PuzzleOutcome.WRONG, currentIndex)
        applySolveRating(solved = false, puzzleRating = puzzles[currentIndex].rating ?: userRating)
        val puzzleId = puzzles[currentIndex].id
        attemptedPuzzleIds.add(puzzleId)
        viewModelScope.launch { progressStore?.markAttempted(puzzleId) }
    }

    private fun applySolveRating(solved: Boolean, puzzleRating: Int) {
        val delta = calculator.change(userRating = userRating, puzzleRating = puzzleRating, solved = solved)
        userRating = ratingStore.apply(delta)
        lastRatingDelta = delta
        val newLevel = RatingLevel(userRating)
        if (newLevel != ratingStore.level()) {
            levelTransition = newLevel
            ratingStore.set(userRating)
        }
    }

    private fun recordOutcome(outcome: PuzzleOutcome, index: Int) {
        if (index in results.indices && results[index] == null) results[index] = outcome
    }

    private fun clearSelection() { selectedSquare = null; attemptedMove = null; hintMove = null }

    private fun orientBoard() { isBoardFlipped = session.userColor == PieceColor.BLACK }

    private fun displayedPosition(): Map<Square, Piece> {
        val base = session.board.pieces
        val attempt = attemptedMove ?: return base
        val piece = base[attempt.from] ?: return base
        return base.toMutableMap().apply { remove(attempt.from); put(attempt.to, piece) }
    }

    private val inReview: Boolean get() = session.state == SOLVED || session.isReviewing
    private val canStepForward: Boolean get() = inReview && session.canStepForward
    private val canStepBack: Boolean get() = inReview && session.canStepBack
    private val hintEnabled: Boolean
        get() = !inReview && (session.state == WAITING_FOR_MOVE || session.state == INCORRECT_MOVE)

    private fun emit() {
        _ui.update {
            it.copy(
                puzzleNumber = currentIndex + 1,
                puzzleCount = puzzles.size,
                displayedPosition = displayedPosition(),
                lastMove = session.lastMove,
                selectedSquare = selectedSquare,
                hintMove = hintMove,
                playerColor = session.userColor,
                isBoardFlipped = isBoardFlipped,
                sessionState = session.state,
                inReview = inReview,
                isReviewing = session.isReviewing,
                canStepForward = canStepForward,
                canStepBack = canStepBack,
                hintEnabled = hintEnabled,
                currentMoveNumber = session.currentMoveNumber,
                totalUserMoves = session.totalUserMoves,
                userRating = userRating,
                lastRatingDelta = lastRatingDelta,
                results = results.toList(),
                levelTransition = levelTransition,
                feedback = feedback(),
            )
        }
    }

    private fun feedback(): TacticsFeedback {
        if (session.state == WAITING_FOR_MOVE) {
            return if (hintMove == null) TacticsFeedback.Instruction("Find the winning move")
            else TacticsFeedback.Instruction("Move the highlighted piece to the marked square")
        }
        return when (session.state) {
            OPPONENT_MOVING -> if (session.isReviewing) TacticsFeedback.OpponentReply else TacticsFeedback.OpponentMoving
            INCORRECT_MOVE -> TacticsFeedback.IncorrectMove
            SOLVED -> if (currentIndex == puzzles.size - 1) TacticsFeedback.TrainingComplete else TacticsFeedback.PuzzleComplete
            WAITING_FOR_MOVE -> TacticsFeedback.Instruction("Find the winning move")
        }
    }

    private fun pickRandom(from: List<Puzzle>, count: Int): List<Puzzle> =
        from.shuffled().take(minOf(count, from.size))

    private companion object {
        const val NEXT_PUZZLE_DELAY_MS = 300L
        const val OPPONENT_REPLY_MS = 450L
        const val WRONG_MOVE_PREVIEW_MS = 550L
    }
}
